jerry-core/ecma/operations/ecma-regexp-object.c

```diff
@@ -238,86 +245,195 @@ ecma_regexp_unicode_advance (const lit_utf8_byte_t **str_p, /**< reference to st
+  JERRY_UNUSED (ctr_obj_p);
+  ecma_object_t *proto_obj_p = ecma_builtin_get (ECMA_BUILTIN_ID_REGEXP_PROTOTYPE);
+#endif /* ENABLED (JERRY_ES2015) */
 
-  ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) object_p;
+  ecma_object_t *new_object_p = ecma_create_object (proto_obj_p,
+                                                    sizeof (ecma_extended_object_t),
+                                                    ECMA_OBJECT_TYPE_CLASS);
 
-  /* Set the internal [[Class]] property */
-  ext_object_p->u.class_prop.class_id = LIT_MAGIC_STRING_REGEXP_UL;
+#if ENABLED (JERRY_ES2015)
+  ecma_deref_object (proto_obj_p);
+#endif /* ENABLED (JERRY_ES2015) */
+
+  ecma_extended_object_t *regexp_obj_p = (ecma_extended_object_t *) new_object_p;
+  regexp_obj_p->u.class_prop.u.value = JMEM_CP_NULL;
+  regexp_obj_p->u.class_prop.class_id = LIT_MAGIC_STRING_REGEXP_MATCHER;
 
-  /* Set bytecode internal property. */
-  ECMA_SET_INTERNAL_VALUE_POINTER (ext_object_p->u.class_prop.u.value, bytecode_p);
-  ecma_bytecode_ref ((ecma_compiled_code_t *) bytecode_p);
+  ecma_value_t status = ecma_builtin_helper_def_prop (new_object_p,
+                                                      ecma_get_magic_string (LIT_MAGIC_STRING_LASTINDEX_UL),
+                                                      ecma_make_uint32_value (0),
+                                                      ECMA_PROPERTY_FLAG_WRITABLE | ECMA_PROP_IS_THROW);
 
-  /* Create and initialize RegExp object properties */
-  ecma_regexp_create_props (object_p);
-  ecma_regexp_initialize_props (object_p,
-                                ecma_get_string_from_value (bytecode_p->source),
-                                bytecode_p->header.status_flags);
+  JERRY_ASSERT (ecma_is_value_true (status));
 
-  return ecma_make_object_value (object_p);
-} /* ecma_op_create_regexp_object_from_bytecode */
+  return new_object_p;
+} /* ecma_op_regexp_alloc */
 
 /**
- * RegExp object creation operation.
+ * RegExpInitialize method
  *
- * See also: ECMA-262 v5, 15.10.4.1
+ * See also: ECMA-262 v6, 21.2.3.2.2
  *
- * @return constructed RegExp object - if pattern and flags were parsed successfully
- *         error ecma value          - otherwise
+ * Note:
+ *      Returned value must be freed with ecma_free_value.
  *
- *         Returned value must be freed with ecma_free_value
+ * @return ecma_value_t
  */
 ecma_value_t
-ecma_op_create_regexp_object (ecma_string_t *pattern_p, /**< input pattern */
-                              uint16_t flags) /**< flags */
+ecma_op_regexp_initialize (ecma_object_t *obj_p, /**< RegExp object */
+                           ecma_value_t pattern_value, /**< pattern of the RegExp object */
+                           ecma_value_t flags_value) /**< flags of the RegExp object */
 {
-  JERRY_ASSERT (pattern_p != NULL);
+  ecma_string_t *pattern_str_p = NULL;
+  uint16_t flags = 0;
+  const re_compiled_code_t *bc_p;
+
+  if (!ecma_is_value_empty (flags_value))
```

Comment:
I don't really like this approach of creating a jack-of-all-trades function to initialize regexp instances. I especially don't like the fact that the bytecode is occasionally passed not-so-transparently as the pattern value.

I'd much prefer if we split this code up into logical parts, and not jumble it together. We could have separate functions to handle regexp allocation/construction, bytecode creation, and regexp initialization, and combine these as needed into two functions that create a regexp instance from a pattern or a bytecode respectively.